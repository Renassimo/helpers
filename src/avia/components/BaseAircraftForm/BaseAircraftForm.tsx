import { KeyboardEvent, useEffect } from 'react';

import { Avia } from '@/avia/types/avia';

import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import SearchIcon from '@mui/icons-material/Search';

import useStateValue from '@/common/hooks/useStateValue';

import AviaInput from '@/avia/components/AviaInput';
import AircraftCard from '@/avia/components/AircraftCard';

const BaseAircraftForm = ({
  registration = null,
  aircraftsResult: {
    aircrafts,
    chosenAircraft,
    retreiveAircrafts,
    chooseAircraft,
    clearChosenAircraft,
    loading,
  },
}: {
  registration?: string | null;
  aircraftsResult: Avia.AircraftsResult;
}) => {
  const [aircraftValue, setAircraftValue] = useStateValue('');

  const onSubmit = async () => {
    await retreiveAircrafts(aircraftValue);
  };

  const lengthOk = aircraftValue.length > 2;

  const onEnter = (event: KeyboardEvent) => {
    if (event.key === 'Enter' && lengthOk) onSubmit();
  };

  useEffect(() => {
    if (!aircraftValue && registration) setAircraftValue(registration);
  }, [registration]);

  return (
    <Box>
      {!chosenAircraft && (
        <Box mt={3}>
          <AviaInput
            title="Aircraft registration"
            value={aircraftValue}
            setValue={setAircraftValue}
            disabled={loading}
            onKeyDown={onEnter}
            adornments={
              <IconButton
                size="small"
                onClick={onSubmit}
                disabled={loading || !lengthOk}
                aria-label="Search"
              >
                <SearchIcon />
              </IconButton>
            }
          />
        </Box>
      )}
      {aircrafts && !chosenAircraft && (
        <Box display="flex" overflow="auto">
          {aircrafts.map((aircraft, index) => (
            <Box
              key={aircraft.id + index}
              mr={1}
              ml={index === 0 ? 1 : undefined}
            >
              <AircraftCard
                data={aircraft}
                onClick={() => chooseAircraft(aircraft.id)}
              />
            </Box>
          ))}
        </Box>
      )}
      {chosenAircraft && (
        <AircraftCard
          data={chosenAircraft}
          onClick={clearChosenAircraft}
          chosen
        />
      )}
    </Box>
  );
};

export default BaseAircraftForm;
