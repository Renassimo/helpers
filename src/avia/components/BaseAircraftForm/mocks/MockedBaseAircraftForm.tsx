const MockedBaseAircraftForm = jest.fn(({ registration }) => (
  <div>MockedBaseAircraftForm - {registration}</div>
));

export default MockedBaseAircraftForm;
