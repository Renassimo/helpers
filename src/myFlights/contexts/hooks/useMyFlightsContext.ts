import { useContext } from 'react';

import MyFlightsContext from '@/myFlights/contexts/MyFlightsContext';

const useMyFlightsContext = () => useContext(MyFlightsContext);

export default useMyFlightsContext;
