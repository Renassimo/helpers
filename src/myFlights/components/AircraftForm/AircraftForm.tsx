import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import BaseAircraftForm from '@/avia/components/BaseAircraftForm';

const AircraftForm = ({ registration }: { registration?: string | null }) => {
  const { aircraftsResult } = useMyFlightsContext();

  return (
    <BaseAircraftForm
      registration={registration}
      aircraftsResult={aircraftsResult}
    />
  );
};

export default AircraftForm;
