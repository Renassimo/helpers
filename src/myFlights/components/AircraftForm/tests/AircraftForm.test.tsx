import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import BaseAircraftForm from '@/avia/components/BaseAircraftForm';
import MockedBaseAircraftForm from '@/avia/components/BaseAircraftForm/mocks';

import AircraftForm from '../AircraftForm';

jest.mock('@/myFlights/contexts/hooks/useMyFlightsContext');
jest.mock('@/avia/components/BaseAircraftForm');

describe('AircraftForm', () => {
  beforeEach(() => {
    (BaseAircraftForm as unknown as jest.Mock).mockImplementation(
      MockedBaseAircraftForm
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    // Arange
    (useMyFlightsContext as unknown as jest.Mock).mockImplementation(
      jest.fn(() => ({ aircraftsResult: 'aircraftsResult' }))
    );
    // Act
    const { baseElement } = renderWithTheme(
      <AircraftForm registration="Registration" />
    );
    // Assert
    expect(baseElement).toMatchSnapshot();
    expect(BaseAircraftForm).toBeCalledWith(
      { aircraftsResult: 'aircraftsResult', registration: 'Registration' },
      {}
    );
  });
});
