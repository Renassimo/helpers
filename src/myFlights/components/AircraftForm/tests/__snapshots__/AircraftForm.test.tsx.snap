// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AircraftForm renders successfully 1`] = `
<body>
  <div>
    <div>
      MockedBaseAircraftForm - 
      Registration
    </div>
  </div>
</body>
`;
